finite_approx/gibbs_sampler_lib.py

```diff
@@ -101,3 +103,213 @@ def sample(self, burnin, num_gibbs_draws):
+                                self.alpha/self.k_approx)\
+                                /(self.x_n + self.alpha/self.k_approx)
+                assert (p_z_nk1_cond_Z >= 0) & (p_z_nk1_cond_Z <= 1)
+
+                if self.z[n,k] == 1:
+                    p_znk_cond_z_stay = p_z_nk1_cond_Z
+                    p_znk_cond_z_update = 1 - p_z_nk1_cond_Z
+                else:
+                    p_znk_cond_z_update = p_z_nk1_cond_Z
+                    p_znk_cond_z_stay = 1 - p_z_nk1_cond_Z
+
+                # next compute p(X|Z_new)
+                # equation (8) in Griffiths and Ghahramani
+
+                # flip z[n,k]
+                self.z[n,k] = 1 - self.z[n,k]
+                [log_lh_X_update, var_inv_update] = update_x_lp_cond_z(\
+                    self.x, self.x2, self.z2, self.z, self.var_inv, self.sigma_eps, \
+                    self.sigma_a, self.k_approx, n, k)
+
+                log_p_znk_cond_z_update = np.log(p_znk_cond_z_update)
+                log_p_zn_update = log_lh_X_update + log_p_znk_cond_z_update \
+                    - sp.misc.logsumexp([log_lh_X_update + log_p_znk_cond_z_update,\
+                        self.log_lh_X+ np.log(p_znk_cond_z_stay)])
+
+                #assert (np.exp(log_p_zn_update) >= 0) & (np.exp(log_p_zn_update) <= 1)
+
+                choice = np.random.binomial(1, np.exp(log_p_zn_update))
+
+                # update Z and prep log likelihood for next round:
+                if choice == 1: # update the necessary quantities
+                    self.log_lh_X = deepcopy(log_lh_X_update)
+                    self.var_inv = deepcopy(var_inv_update)
+                    self.z2 = np.dot(self.z.T, self.z)
+                else:
+                    # flip back to the orginal value of z[n,k]
+                    self.z[n,k] = 1 - self.z[n,k]
+
+        if keep_draw:
+            self.z_draws.append(self.z)
+
+    def sample(self, burnin, num_gibbs_draws):
+
+        print('Sampling:')
+        bar = progressbar.ProgressBar(max_value=num_gibbs_draws + burnin)
+        for n in bar(range(num_gibbs_draws + burnin)):
+            self.draw_z(keep_draw = n >= burnin)
+
+        print('Done sampling :)')
+
+
+def update_x_lp_cond_z(x, x2, z2, z_update, var_inv, sigma_eps, sigma_a, \
+                        K_approx, n,k):
+# likelihood p(X|Z)-- equation (8) in Griffiths and Ghahramani
+# http://mlg.eng.cam.ac.uk/zoubin/papers/ibp-nips05.pdf
+# outputs the likelihood at Z  when Z has component n,k flipped
+# var_inv refers to inv(Z^T * Z + sigma_eps/sigma_a I), the previous inverse
+
+    x_d = np.shape(x)[1]
+    x_n = np.shape(x)[0]
+    k_approx = np.shape(z_update)[1]
+
+    assert np.shape(x)[0] == np.shape(z_update)[0]
+
+    inv_var_flip = update_inv_var(z2, z_update, var_inv, sigma_eps, sigma_a, n, k)
+
+    mean_a = np.dot(np.dot(inv_var_flip, z_update.T), x)
+
+    log_likelihood = -1/(2*sigma_eps) * \
+            np.trace(x2 - np.dot(x.T, np.dot(z_update, mean_a)) )
```

Comment:
z_update is the whole matrix, even though you have only changed one thing from last time.  This is costly.  This goes here and everywhere else you calculate Z' Z or X' Z inside the loop in draw_z.

Think about it this way -- look through the code and see where you're doing floating point operations.  That, and memory allocation, are what's taking time.  Each inner product like this takes ~ n * k floating point operations.  Updating a single row of Z takes ~ k floating point operations.  That is, proportionally on the order of (n - 1) / n of your floating point operations are unnecessary.  Eliminating this kind of waste could speed the algorithm up quite markedly.

You can write a function that only updates the term you need, and then write a simple version like this to test that it does the right thing.